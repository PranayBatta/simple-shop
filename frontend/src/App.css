@import '~antd/dist/antd.css';

.formInput {
  color: white;
}

.ant-card-meta-description {
  height: 110px;
  overflow: scroll;
}

.ant-card-actions {
  background-color: white;
}

.shop-header {
  height: 6em;
}

.art-manager-list-container {
  padding:1.5em 2em;
  background-color: white;
  height: 82vh;
}

.ant-list-item-action li {
  cursor: default;
}

.ant-modal-content {
  box-shadow: 0 0 5px rgb(171, 177, 186);
}

.ant-drawer-header {
  height: 6em;
  padding: 2em;
}

.ant-drawer-title {
  font-size: 1.5em;
}

.app-footer {
  background-color: #001529;
  text-align: right;
  padding-top: 1.25em;
}

.github-logo {
  height: 2.5em;
  image-rendering: -moz-crisp-edges;         /* Firefox */
  image-rendering:   -o-crisp-edges;         /* Opera */
  image-rendering: -webkit-optimize-contrast;/* Webkit (non-standard naming) */
  image-rendering: crisp-edges;
  -ms-interpolation-mode: nearest-neighbor;  /* IE (non-standard property) */
}

.shop-list {
  max-height: 60vh;
}

.shop-sider {
  max-height: 82vh;
  overflow: scroll;
}

.shop-list-container {
  background-color: white;
  padding-top: 3em;
}