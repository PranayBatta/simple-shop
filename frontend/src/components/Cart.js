import React from 'react';
import moment from 'moment';
import axios from 'axios';
import { ArrowUpOutlined, ArrowDownOutlined } from '@ant-design/icons';
import {
  Button,
  Statistic,
  Card,
  Row,
  Col 
} from 'antd';

const layout = {
  labelCol: { span: 8 },
  wrapperCol: { span: 16 },
};

const Cart = (props) => {
//   const [form] = Form.useForm();
  
//   const onFinish = async (values) => {
//     const { artist, date, description, height, name, width } = values;
//     let art = {
//       artist: artist ? artist : 'Hughie Devore',
//       date: date ? date.format('YYYY-MM-DD') : moment().format('YYYY-MM-DD'),
//       description: description ? description : 'A beautiful view of snowy peaks in the rocky mountains.',
//       height: height ? height : 1000,
//       name: name ? name : 'Rocky Mountain Masterpiece',
//       width: width ? width : 1250,
//     };

//     const response = await axios.post('http://localhost:3100/art', art);
//     const body = response.data;
//     if (response.status !== 201) {
//       notification.error({
//         message: 'Error, your art could not be added to the manager.',
//         description: body.message,
//         onClick: () => {
//           console.log('Notification Clicked!');
//         },
//       });
//     } else {
//       notification.success({
//         message: 'Your art was added to the manager!',
//         description: body.message,
//         onClick: () => {
//           console.log('Notification Clicked!');
//         },
//       });
//       props.getArtList();
//     }
//   };

  return (
    <div className="site-statistic-demo-card">
        <Card>
            <Statistic
                title="Subtotal"
                value={100.57}
                precision={2}
                valueStyle={{ color: "#3f8600" }}
                prefix="$"
            />
        </Card>
        <Card>
            <Statistic
                title="Shipping & Taxes"
                value={21.00}
                precision={2}
                valueStyle={{ color: "#cf1322" }}
                prefix="$"
            />
        </Card>
        <Card>
            <Statistic
                title="Amount Due"
                value={121.57}
                precision={2}
                valueStyle={{ color: "black", fontWeight: 800 }}
                prefix="$"
            />
        </Card>
    </div>
  );
}

export default Cart;