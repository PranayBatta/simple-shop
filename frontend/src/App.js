import React, { Component } from 'react';
import axios from 'axios';
import styled from 'styled-components';
import './App.css';
import {
  Button,
  Card,
  Layout,
  List,
  notification,
  Popconfirm,
} from 'antd';
import Cart from './components/Cart';
import { DeleteOutlined, EditOutlined, SettingOutlined, EllipsisOutlined } from '@ant-design/icons';
import GithubLogo from './GitHub-Mark-Light-64px.png';

const { Content, Footer, Header, Sider } = Layout;
const { Item } = List;
const { Meta } = Card;

const PageHeader = styled.h1`
  color: white;
  font-size: 3em;
  padding-top: .2em;
  font-weight: 700;
`;

const Price = styled.span`
  color: black;
  font-weight: 600;
  font-size: 1.25em;
  line-height: 2.25em;
`;

class App extends Component {
  state = {
    cartId: 0,
    cartList: [],
    productList: [],
    visible: false,
  };

  createCart = async () => {
    const response = await axios.post('http://localhost:3100/carts');
    const body = response.data;
    if (response.status !== 201) {
      throw Error(body.message);
    }
    this.setState({ 
      cartID: body[0].id,
    });
    // Persist the cartId in the client with local storage
    localStorage.setItem('cartId', body[0].id);
  }

  getCartList = async () => {
    let cartListBeta = [];
    let cartList = [];
    const cartId = localStorage.getItem('cartId');
    const response = await axios.get(`http://localhost:3100/carts/${cartId}`);
    const body = response.data;
    if (response.status !== 200) {
      throw Error(body.message);
    }
    const productAggregate = body.reduce(function(results, item) {
      (results[item.product_id] = results[item.product_id] || []).push(item);
      return results;
    }, {});
    Object.keys(productAggregate).forEach(item => {
      let listItem = {};
      listItem.id = item;
      listItem.quantity = productAggregate[item].length;
      cartListBeta.push(listItem);
    });
    cartList = cartListBeta.map(async item => {
      await this.getProductById(item.id)
      .then(product => {
        item.price = product[0].price;
        item.name = product[0].name;
        return item;
      });
    });
    if (cartList) {
      this.setState({ 
        cartList: cartList,
      });
    }
  }

  getProductList = async () => {
    const response = await axios.get('http://localhost:3100/products');
    const body = response.data;
    if (response.status !== 200) {
      throw Error(body.message) 
    }
    this.setState({ 
      productList: body,
    });
  }

  getProductById = async (id) => {
    const response = await axios.get(`http://localhost:3100/products/${id}`);
    const body = response.data;
    if (response.status !== 200) {
      throw Error(body.message);
    }
    return body;
  };

  addToCart = async (e) => {
    const productId = e.target.getAttribute('id');
    const response = await axios.post(`http://localhost:3100/carts/${this.state.cartId}`, {product_id: productId});
    const body = response.data;
    if (response.status !== 201) {
      notification.error({
        message: 'Error, the selected item could not be added to your cart.',
        description: body.message,
      });
    } else {
      notification.success({
        message: 'Item added to your cart!',
        description: body.message,
      });
      this.getCartList();
    }
  }

  // TODO: Remove Product from Cart

  // deleteArt = async (id) => {
  //   console.log(id);
  //   const response = await axios.delete(`http://localhost:3100/art/${id}`);
  //   const body = response.data;
  //   if (response.status !== 200) {
  //     notification.error({
  //       message: 'Error, your art could not be deleted.',
  //       description: body.message,
  //     });
  //   } else {
  //     notification.success({
  //       message: 'Your art was deleted from sthe manager!',
  //       description: body.message,
  //     });
  //     this.getArtList();
  //   }
  // }

  // TODO: Add Cart button in header and use drawer for cart
  showDrawer = () => {
    this.setState({visible: true});
  };

  handleClose = e => {
    this.setState({
      visible: false,
    });
  };

  componentDidMount() {
    const cartId = localStorage.getItem('cartId');
    if (cartId !== 0) {
      this.setState({
        cartId: cartId
      });
    }
    this.getProductList();
    this.getCartList();
  }

  render() {
    const { productList } = this.state;    
    return (
      <Layout style={{minHeight: '100vh' }}>
        <Header className="shop-header">
          <PageHeader>Simple Shop</PageHeader>
        </Header>
        <Layout>
          <Content
            className="shop-list-container"
          >
            <List 
              className="shop-list"
              grid={{
                gutter: [24, 40],
                sm: 1,
                md: 2,
                lg: 2,
                xl: 2,
                xxl: 3,
              }}
              dataSource={productList}
              size="large"
              split={true}
              renderItem={item => {   
                return (
                  <Item>
                    <Card
                      hoverable
                      style={{ width: '100%'}}
                      onClick={this.addToCart}
                      cover={
                        <img
                          alt="example"
                          src={item.image_url}
                        />
                      }
                      actions={[
                        <Price>{item.price}</Price>,
                        <Button size="large" type="primary" id={item.id}>Add to Cart</Button>,
                      ]}
                    >
                      <Meta
                        title={item.name}
                        description={item.description}
                      />
                    </Card>
                  </Item>
                );
              }}
            />
          </Content>
          <Sider
            width={450}
            height="82vh"
            style={{
              padding: '2em',
              backgroundColor: 'rgb(236, 236, 236)',
            }}
            breakpoint="lg"
            collapseWidth="0"
            className="shop-sider"
          >
            <Cart cartId={this.state.cartId} cartList={this.state.cartList} getCartList={this.getCartList}/>
          </Sider>
        </Layout>
        <Footer className="app-footer">
          <a href="https://github.com/hughdevore/simple-shop">
            <img src={GithubLogo} className="github-logo" alt="GitHub Logo with link to repository"/>
          </a>
        </Footer>
      </Layout>
    );
  }
}
export default App;