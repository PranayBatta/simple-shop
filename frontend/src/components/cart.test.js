import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import renderer from 'react-test-renderer';
import Cart from "./Cart";
import { convertToMoney } from '../App';
jest.mock('../App');

jest.mock('rc-util/lib/Portal');

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('The cart component.', () => {
  it("renders without any products if the list or id are empty.", () => {
    act(() => {
      render(<Cart convertToMoney={convertToMoney} />, container);
    });
    expect(container.textContent).toBe("Cart Summary0 products in your cartAdd items to your cart!Taxes$NaNCart Total$NaNCheckout");
  });

  it('renders the cart component correctly.', () => {
    const tree = renderer
      .create(<Cart convertToMoney={convertToMoney} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});