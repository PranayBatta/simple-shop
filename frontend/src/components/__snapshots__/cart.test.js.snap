// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The cart component. renders the cart component correctly. 1`] = `
<div>
  <div
    style={
      Object {
        "borderBottom": "1px solid rgb(82, 82, 82)",
      }
    }
  >
    <h1
      className="sc-AxiKw dgGdNC"
    >
      Cart Summary
    </h1>
    <span
      className="sc-AxhCb gWooew"
    >
      0 products in your cart
    </span>
  </div>
  <div
    className="ant-list ant-list-split"
  >
    <div
      className="ant-spin-nested-loading"
    >
      <div
        className="ant-spin-container"
      >
        <div
          className="ant-list-empty-text"
        >
          <div
            className="ant-empty"
            style={
              Object {
                "fontSize": "7em",
              }
            }
          >
            <div
              className="ant-empty-image"
            >
              <span
                aria-label="shopping-cart"
                className="anticon anticon-shopping-cart"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  className=""
                  data-icon="shopping-cart"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="0 0 1024 1024"
                  width="1em"
                >
                  <path
                    d="M922.9 701.9H327.4l29.9-60.9 496.8-.9c16.8 0 31.2-12 34.2-28.6l68.8-385.1c1.8-10.1-.9-20.5-7.5-28.4a34.99 34.99 0 00-26.6-12.5l-632-2.1-5.4-25.4c-3.4-16.2-18-28-34.6-28H96.5a35.3 35.3 0 100 70.6h125.9L246 312.8l58.1 281.3-74.8 122.1a34.96 34.96 0 00-3 36.8c6 11.9 18.1 19.4 31.5 19.4h62.8a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7h161.1a102.43 102.43 0 00-20.6 61.7c0 56.6 46 102.6 102.6 102.6s102.6-46 102.6-102.6c0-22.3-7.4-44-20.6-61.7H923c19.4 0 35.3-15.8 35.3-35.3a35.42 35.42 0 00-35.4-35.2zM305.7 253l575.8 1.9-56.4 315.8-452.3.8L305.7 253zm96.9 612.7c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6zm325.1 0c-17.4 0-31.6-14.2-31.6-31.6 0-17.4 14.2-31.6 31.6-31.6s31.6 14.2 31.6 31.6a31.6 31.6 0 01-31.6 31.6z"
                  />
                </svg>
              </span>
            </div>
            <p
              className="ant-empty-description"
            >
              <span
                style={
                  Object {
                    "fontSize": "18px",
                  }
                }
              >
                Add items to your cart!
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    className="sc-AxheI fxScyK"
  >
    <span>
      Taxes
    </span>
    <span
      className="sc-AxjAm gkIcmJ"
    >
      $NaN
    </span>
  </span>
  <span
    className="sc-Axmtr RqTAG"
  >
    <span
      style={
        Object {
          "color": "black",
        }
      }
    >
      Cart Total
    </span>
    <span
      className="sc-AxirZ hCljwO"
    >
      $NaN
    </span>
  </span>
  <div
    className="sc-AxhUy fUUiqO"
  >
    <button
      className="ant-btn checkout-button ant-btn-primary ant-btn-lg"
      onClick={[Function]}
      type="button"
    >
      <span>
        Checkout
      </span>
    </button>
  </div>
</div>
`;
